File: adapters/runtimes/java/src/bridge.ts
/**
 * File: adapters/runtimes/java/src/bridge.ts
 * Purpose: Invokes the dependency-free Java SDK JSON bridge through the shared shell-free process boundary and normalizes responses for the TypeScript control plane.
 * Author: Raushan Raj
 */
import type { EvidenceRecord } from '../../../../packages/contracts/src/evidence';
import type { RuntimeIdentity } from '../../../../packages/contracts/src/execution';
import type { UniversalTestResult } from '../../../../packages/contracts/src/result';
import { probeProcess, type ProcessProbeOptions, type ProcessProbeResult } from '../../../../packages/runtime/src';

export interface JavaBridgeClientOptions {
  classpath: string;
  executable?: string;
  mainClass?: string;
  timeoutMs?: number;
  cwd?: string;
  environment?: Record<string, string | undefined>;
  probe?: (
    command: string,
    args?: string[],
    options?: ProcessProbeOptions,
  ) => Promise<ProcessProbeResult>;
}

interface JavaBridgeEnvelope<T> {
  ok: boolean;
  data?: T;
  error?: { type?: string; message?: string };
}

export class JavaBridgeClient {
  private readonly executable: string;
  private readonly mainClass: string;
  private readonly timeoutMs: number;
  private readonly probe: NonNullable<JavaBridgeClientOptions['probe']>;

  constructor(private readonly options: JavaBridgeClientOptions) {
    if (!options.classpath.trim()) throw new Error('Java bridge classpath is required.');
    this.executable = options.executable ?? process.env.QUALYNTRA_JAVA_EXECUTABLE ?? 'java';
    this.mainClass = options.mainClass ?? 'io.qualyntra.sdk.Bridge';
    this.timeoutMs = options.timeoutMs ?? 30_000;
    this.probe = options.probe ?? probeProcess;
  }

  async health(): Promise<Record<string, unknown>> {
    return await this.invoke<Record<string, unknown>>({ operation: 'health' });
  }

  async normalizeJUnit(
    xml: string,
    runId: string,
    runtime: RuntimeIdentity,
  ): Promise<UniversalTestResult[]> {
    const response = await this.invoke<{ results: UniversalTestResult[] }>({
      operation: 'results.junit.normalize',
      xml,
      runId,
      runtime,
    });
    return Array.isArray(response.results) ? response.results : [];
  }

  async normalizeTestNg(
    xml: string,
    runId: string,
    runtime: RuntimeIdentity,
  ): Promise<UniversalTestResult[]> {
    const response = await this.invoke<{ results: UniversalTestResult[] }>({
      operation: 'results.testng.normalize',
      xml,
      runId,
      runtime,
    });
    return Array.isArray(response.results) ? response.results : [];
  }

  async collectFileEvidence(input: {
    runId: string;
    path: string;
    kind?: string;
    contentType?: string;
  }): Promise<EvidenceRecord> {
    const response = await this.invoke<{ evidence: EvidenceRecord }>({
      operation: 'evidence.file',
      runId: input.runId,
      path: input.path,
      kind: input.kind ?? 'custom',
      contentType: input.contentType ?? 'application/octet-stream',
    });
    if (!response.evidence) throw new Error('Java bridge returned no evidence record.');
    return response.evidence;
  }

  private async invoke<T>(payload: Record<string, unknown>): Promise<T> {
    const result = await this.probe(
      this.executable,
      ['-cp', this.options.classpath, this.mainClass],
      {
        cwd: this.options.cwd,
        env: this.options.environment,
        timeoutMs: this.timeoutMs,
        input: JSON.stringify(payload),
      },
    );

    if (!result.available) throw new Error('Java bridge executable is unavailable.');
    if (result.timedOut) throw new Error('Java bridge request timed out.');

    let envelope: JavaBridgeEnvelope<T>;
    try {
      envelope = JSON.parse(result.stdout.trim()) as JavaBridgeEnvelope<T>;
    } catch {
      throw new Error('Java bridge returned malformed JSON.');
    }

    if (result.exitCode !== 0 || !envelope.ok) {
      const type = envelope.error?.type ?? 'JavaBridgeError';
      const message = envelope.error?.message ?? 'Java bridge request failed.';
      throw new Error(`${type}: ${message}`);
    }
    if (envelope.data === undefined) throw new Error('Java bridge returned no data.');
    return envelope.data;
  }
}

File: sdks/java/src/main/java/io/qualyntra/sdk/Bridge.java
/**
 * File: sdks/java/src/main/java/io/qualyntra/sdk/Bridge.java
 * Purpose: Exposes the Java SDK through a stable dependency-free JSON-over-stdio bridge for runtime health, result normalization, and evidence collection.
 * Author: Raushan Raj
 */
package io.qualyntra.sdk;

import java.nio.charset.StandardCharsets;
import java.nio.file.Path;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

public final class Bridge {
    private Bridge() {}

    public static Map<String, Object> handle(Map<String, Object> payload) throws Exception {
        String operation = string(payload.get("operation"));
        return switch (operation) {
            case "health" -> Map.of("status", "healthy", "runtime", RuntimeProbe.discover().toMap());
            case "results.junit.normalize" -> Map.of(
                "results",
                ResultNormalizer.parseJUnit(
                    requiredString(payload, "xml"),
                    requiredString(payload, "runId"),
                    object(payload.get("runtime"))
                )
            );
            case "results.testng.normalize" -> Map.of(
                "results",
                ResultNormalizer.parseTestNg(
                    requiredString(payload, "xml"),
                    requiredString(payload, "runId"),
                    object(payload.get("runtime"))
                )
            );
            case "evidence.file" -> Map.of(
                "evidence",
                Evidence.fromFile(
                    requiredString(payload, "runId"),
                    stringOr(payload.get("kind"), "custom"),
                    Path.of(requiredString(payload, "path")),
                    stringOr(payload.get("contentType"), "application/octet-stream")
                ).toMap()
            );
            default -> throw new IllegalArgumentException("Unsupported bridge operation: " + operation);
        };
    }

    public static void main(String[] args) {
        Map<String, Object> response = new LinkedHashMap<>();
        try {
            String input = new String(System.in.readAllBytes(), StandardCharsets.UTF_8);
            Map<String, Object> payload = Json.parseObject(input.isBlank() ? "{}" : input);
            response.put("ok", true);
            response.put("data", handle(payload));
            System.out.println(Json.stringify(response));
        } catch (Exception exception) {
            response.put("ok", false);
            response.put("error", Map.of(
                "type", exception.getClass().getSimpleName(),
                "message", exception.getMessage() == null ? "Java bridge failure" : exception.getMessage()
            ));
            System.out.println(Json.stringify(response));
            System.exit(1);
        }
    }

    @SuppressWarnings("unchecked")
    private static Map<String, Object> object(Object value) {
        return value instanceof Map<?, ?> map ? (Map<String, Object>) map : Map.of("language", "java", "runner", "unknown");
    }

    private static String requiredString(Map<String, Object> payload, String key) {
        String value = string(payload.get(key));
        if (value == null || value.isBlank()) throw new IllegalArgumentException(key + " is required");
        return value;
    }

    private static String string(Object value) {
        return value == null ? null : String.valueOf(value);
    }

    private static String stringOr(Object value, String fallback) {
        String text = string(value);
        return text == null || text.isBlank() ? fallback : text;
    }
}
